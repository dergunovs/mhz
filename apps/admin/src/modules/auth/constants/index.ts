export const API_POST_LOGIN = 'auth/login';
export const API_GET_CHECK_AUTH = 'auth/check';

export const URL_LOGIN = '/';
