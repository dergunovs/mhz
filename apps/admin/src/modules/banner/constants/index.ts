export const URL_BANNER = '/banners';
export const URL_BANNER_CREATE = `${URL_BANNER}/create`;
export const URL_BANNER_EDIT = `${URL_BANNER}/edit`;
