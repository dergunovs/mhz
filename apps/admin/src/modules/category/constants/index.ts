export const URL_CATEGORY = '/categories';
export const URL_CATEGORY_CREATE = `${URL_CATEGORY}/create`;
export const URL_CATEGORY_EDIT = `${URL_CATEGORY}/edit`;

export const CATEGORY_FIELD_TYPE_OPTIONS = ['string', 'number', 'boolean'];
