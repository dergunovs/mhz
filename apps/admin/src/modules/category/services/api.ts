import { ComputedRef } from 'vue';

import { api } from 'mhz-helpers';
import { API_CATEGORY, ICategory, IBaseReply } from 'mhz-contracts';

export async function getCategoriesApi() {
  const { data } = await api.get<{ data: ICategory[] }>(API_CATEGORY);

  return data.data;
}

export async function getCategoryApi(id?: ComputedRef<string | string[]>) {
  if (!id?.value) return null;

  const { data } = await api.get<{ data: ICategory }>(`${API_CATEGORY}/${id.value}`);

  return data.data;
}

export async function postCategoryApi(formData: ICategory) {
  const { data } = await api.post<IBaseReply>(API_CATEGORY, formData);

  return data;
}

export async function updateCategoryApi(formData: ICategory) {
  const { data } = await api.patch<IBaseReply>(`${API_CATEGORY}/${formData._id}`, formData);

  return data;
}

export async function deleteCategoryApi(id?: string) {
  if (!id) return null;

  const { data } = await api.delete<IBaseReply>(`${API_CATEGORY}/${id}`);

  return data;
}
