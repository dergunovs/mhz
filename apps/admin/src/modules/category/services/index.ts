import { ComputedRef } from 'vue';

import { api, useQuery, useMutation } from 'mhz-helpers';
import { API_CATEGORY, ICategory, IBaseReply } from 'mhz-contracts';

export function getCategories(options?: object) {
  async function fn() {
    const { data } = await api.get<{ data: ICategory[] }>(API_CATEGORY);

    return data.data;
  }

  return useQuery({ queryKey: [API_CATEGORY], queryFn: fn, ...options });
}

export function getCategory(id?: ComputedRef<string | string[]>) {
  async function fn() {
    if (!id?.value) return null;

    const { data } = await api.get<{ data: ICategory }>(`${API_CATEGORY}/${id.value}`);

    return data.data;
  }

  return useQuery({ queryKey: [API_CATEGORY, id], queryFn: fn });
}

export function postCategory(options: object) {
  async function fn(formData: ICategory) {
    const { data } = await api.post<IBaseReply>(API_CATEGORY, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}

export function updateCategory(options: object) {
  async function fn(formData: ICategory) {
    const { data } = await api.patch<IBaseReply>(`${API_CATEGORY}/${formData._id}`, formData);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}

export function deleteCategory(options: object) {
  async function fn(id?: string) {
    if (!id) return null;

    const { data } = await api.delete<IBaseReply>(`${API_CATEGORY}/${id}`);

    return data;
  }

  return useMutation({ mutationKey: [API_CATEGORY], mutationFn: fn, ...options });
}
