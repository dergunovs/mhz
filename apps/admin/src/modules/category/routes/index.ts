import { URL_CATEGORY, URL_CATEGORY_CREATE, URL_CATEGORY_EDIT } from '@/category/constants';

export const categoryRoutes = [
  {
    path: URL_CATEGORY,
    name: 'CategoryList',
    component: () => import('@/category/pages/CategoryListPage.vue'),
  },
  {
    path: URL_CATEGORY_CREATE,
    name: 'CategoryCreate',
    component: () => import('@/category/pages/CategoryCreatePage.vue'),
  },
  {
    path: `${URL_CATEGORY_EDIT}/:id`,
    name: 'CategoryEdit',
    component: () => import('@/category/pages/CategoryEditPage.vue'),
  },
];
