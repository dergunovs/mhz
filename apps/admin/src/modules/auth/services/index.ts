import { Ref } from 'vue';
import { useMutation, useQuery } from '@tanstack/vue-query';

import { ILoginFormData } from '@/auth/interface';
import { API_GET_CHECK_AUTH, API_POST_LOGIN } from '@/auth/constants';

import { api } from '@/common/services/api';

export function postLogin(formData: Ref<ILoginFormData>, options: object) {
  async function fn(): Promise<string | undefined> {
    const { data } = await api.post(API_POST_LOGIN, formData.value);

    return data.token;
  }

  return useMutation({
    mutationKey: [API_POST_LOGIN, formData],
    mutationFn: fn,
    ...options,
  });
}

export function getCheckAuth(options: object) {
  async function fn() {
    const { data } = await api.get(API_GET_CHECK_AUTH);

    return data;
  }

  return useQuery({
    queryKey: [API_GET_CHECK_AUTH],
    queryFn: fn,
    ...options,
  });
}
