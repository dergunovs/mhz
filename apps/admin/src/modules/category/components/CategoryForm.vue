<template>
  <form @submit.prevent="props.category?._id ? update() : submit()" :class="$style.form">
    <UiField label="Title" isRequired :error="error('title')">
      <UiInput v-model="formData.title" isFocus />
    </UiField>

    <UiField label="Description" isRequired :error="error('description')">
      <UiEditor v-model="formData.description" />
    </UiField>

    <CategoryFieldList
      v-if="formData.fields?.length"
      :fields="formData.fields"
      :isShowCategoryFieldForm="isShowCategoryFieldForm"
      @edit="editCategoryField"
    />

    <UiButton @click="showCategoryField" v-if="!isShowCategoryFieldForm" layout="secondary">
      Add Category Field
    </UiButton>

    <CategoryFieldForm
      v-if="isShowCategoryFieldForm"
      :categoryField="editableCategoryField"
      @add="addCategoryField"
      @update="updateCategoryField"
      @delete="deleteCategoryField"
      @hide="isShowCategoryFieldForm = false"
    />

    <UiUpload
      label="Upload icon"
      :file="iconFile"
      isSingle
      isRequired
      :isDisabled="!!formData.iconUrl"
      :extensions="['svg']"
      :error="error('iconUrl')"
      @add="addIconFile"
      @remove="removeIconFile"
      @upload="mutateUploadFile(iconFile)"
    />

    <ImagePreview v-if="formData.iconUrl" :urls="[formData.iconUrl]" @delete="formData.iconUrl = ''" />

    <FormButtons :id="props.category?._id" :isLoading="isLoadingPost || isLoadingUpdate" @delete="handleDelete" />
  </form>
</template>

<script setup lang="ts">
import { ref, computed, onMounted } from 'vue';
import { useRouter } from 'vue-router';

import { UiField, UiInput, UiButton, UiUpload, toast, UiEditor } from 'mhz-ui';
import { clone, deleteTempId, useValidator, required, useQueryClient } from 'mhz-helpers';
import { API_CATEGORY, ICategory, ICategoryField } from 'mhz-contracts';

import CategoryFieldForm from '@/category/components/CategoryFieldForm.vue';
import CategoryFieldList from '@/category/components/CategoryFieldList.vue';
import ImagePreview from '@/common/components/ImagePreview.vue';
import FormButtons from '@/common/components/FormButtons.vue';

import { postCategory, updateCategory, deleteCategory } from '@/category/services';
import { uploadFile } from '@/common/services';
import { URL_CATEGORY } from '@/category/constants';

interface IProps {
  category?: ICategory;
}

const props = defineProps<IProps>();

const queryClient = useQueryClient();

const router = useRouter();

const formData = ref<ICategory>({
  title: '',
  description: '',
  iconUrl: '',
  fields: [],
});

const isShowCategoryFieldForm = ref(false);

const editableCategoryField = ref<ICategoryField>();

const { mutate: mutatePost, isPending: isLoadingPost } = postCategory({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_CATEGORY] });
    toast.success('Category added');
    router.push(URL_CATEGORY);
  },
});

const { mutate: mutateUpdate, isPending: isLoadingUpdate } = updateCategory({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_CATEGORY] });
    toast.success('Category updated');
  },
});

const { mutate: mutateDelete } = deleteCategory({
  onSuccess: async () => {
    queryClient.removeQueries({ queryKey: [API_CATEGORY] });
    await queryClient.refetchQueries({ queryKey: [API_CATEGORY] });
    toast.success('Category deleted');
    router.push(URL_CATEGORY);
  },
});

const rules = computed(() => {
  return {
    title: required,
    description: required,
    iconUrl: required,
  };
});

const { error, isValid } = useValidator(formData, rules);

function showCategoryField() {
  editableCategoryField.value = undefined;
  isShowCategoryFieldForm.value = true;
}

function editCategoryField(field: ICategoryField) {
  editableCategoryField.value = field;
  isShowCategoryFieldForm.value = true;
}

function addCategoryField(field: ICategoryField) {
  formData.value.fields?.push(field);
}

function updateCategoryField(fieldToUpdate: ICategoryField) {
  formData.value.fields = formData.value.fields?.map((field) =>
    field._id === fieldToUpdate._id ? fieldToUpdate : field
  );
}

function deleteCategoryField(fieldId: string) {
  formData.value.fields = formData.value.fields?.filter((field) => field._id !== fieldId);
}

function submit() {
  if (isValid()) {
    if (formData.value.fields) formData.value.fields = deleteTempId(formData.value.fields);
    mutatePost(formData.value);
  }
}

function update() {
  if (isValid()) {
    if (formData.value.fields) formData.value.fields = deleteTempId(formData.value.fields);
    mutateUpdate(formData.value);
  }
}

function handleDelete(id: string) {
  mutateDelete(id);
}

const iconFile = ref<File>();

function addIconFile(file: File) {
  iconFile.value = file;
}

function removeIconFile() {
  iconFile.value = undefined;
}

const { mutate: mutateUploadFile } = uploadFile({
  onSuccess: (data: string) => {
    formData.value.iconUrl = data;
    removeIconFile();
    toast.success('Icon added');
  },
});

onMounted(() => {
  if (props.category) formData.value = clone(props.category);
});
</script>

<style module lang="scss">
.form {
  display: flex;
  flex-direction: column;
  gap: 24px;
  align-items: flex-start;
}
</style>
