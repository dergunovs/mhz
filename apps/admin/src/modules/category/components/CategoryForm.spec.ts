import { DefineComponent } from 'vue';
import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { VueWrapper, enableAutoUnmount, flushPromises } from '@vue/test-utils';
import { ICategory, ICategoryField, IBaseReply, API_CATEGORY } from 'mhz-contracts';
import { toast } from 'mhz-ui';
import * as helpers from 'mhz-helpers';
import { dataTest } from 'mhz-helpers';

import CategoryForm from './CategoryForm.vue';

import { wrapperFactory, mockMutationReply, router } from '@/common/test';
import { CATEGORY } from '@/category/fixtures';
import { URL_CATEGORY } from '@/category/constants';
import * as categoryServices from '@/category/services';

let onSuccessPost: () => void;
const spyMutatePost = vi.fn();

vi.spyOn(categoryServices, 'postCategory').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessPost = options.onSuccess;

  return mockMutationReply<IBaseReply, ICategory>(spyMutatePost);
});

let onSuccessUpdate: () => void;
const spyMutateUpdate = vi.fn();

vi.spyOn(categoryServices, 'updateCategory').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessUpdate = options.onSuccess;

  return mockMutationReply<IBaseReply, ICategory>(spyMutateUpdate);
});

let onSuccessDelete: () => void;
const spyMutateDelete = vi.fn();

vi.spyOn(categoryServices, 'deleteCategory').mockImplementation((options: { onSuccess?: () => void }) => {
  if (options.onSuccess) onSuccessDelete = options.onSuccess;

  return mockMutationReply<IBaseReply, string | undefined>(spyMutateDelete);
});

const spyRefetchQueries = vi.fn();
const spyRemoveQueries = vi.fn();

vi.spyOn(helpers, 'useQueryClient').mockReturnValue({
  refetchQueries: spyRefetchQueries,
  removeQueries: spyRemoveQueries,
} as unknown as helpers.QueryClient);

const spyRouterPush = vi.spyOn(router, 'push');
const spyToastSuccess = vi.spyOn(toast, 'success');

const QUERY_KEY = { queryKey: [API_CATEGORY] };

const categoryForm = dataTest('category-form');
const categoryFormTitle = dataTest('category-form-title');
const categoryFormDescription = dataTest('category-form-description');
const categoryFormFieldList = dataTest('category-form-field-list');
const categoryFormAddCategory = dataTest('category-form-add-category');
const categoryFormFieldForm = dataTest('category-form-field-form');
const categoryFormIcon = dataTest('category-form-icon');
const categoryFormButtons = dataTest('category-form-buttons');

let wrapper: VueWrapper;

beforeEach(() => {
  wrapper = wrapperFactory(CategoryForm, {
    props: {
      category: CATEGORY.data,
    },
  });
});

enableAutoUnmount(afterEach);

describe('CategoryForm', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(categoryForm)).toBeTruthy();
  });

  it('shows category content in fields if props is passed', async () => {
    expect(wrapper.findComponent(categoryFormTitle).attributes('modelvalue')).toBe(CATEGORY.data.title);
    expect(wrapper.findComponent(categoryFormDescription).attributes('modelvalue')).toBe(CATEGORY.data.description);
    expect(wrapper.findComponent(categoryFormIcon).attributes('urls')).toBe(CATEGORY.data.iconUrl);

    expect(
      wrapper.findComponent<DefineComponent<{ fields: ICategoryField[] }>>(categoryFormFieldList).vm.$props.fields
    ).toEqual(CATEGORY.data.fields);
  });

  it('shows icon preview if category has fields', async () => {
    expect(wrapper.findComponent(categoryFormIcon).exists()).toBe(true);

    await wrapper.setProps({ category: undefined });

    expect(wrapper.findComponent(categoryFormIcon).exists()).toBe(true);
  });

  it('shows field list if category has fields', async () => {
    expect(wrapper.findComponent(categoryFormFieldList).exists()).toBe(true);

    await wrapper.setProps({ category: undefined });

    expect(wrapper.findComponent(categoryFormFieldList).exists()).toBe(true);
  });

  it('shows field form by edit form emit', async () => {
    const FIELD_TO_EDIT = CATEGORY.data.fields?.[0];

    expect(wrapper.findComponent(categoryFormFieldForm).exists()).toBe(false);

    await wrapper.findComponent<DefineComponent>(categoryFormFieldList).vm.$emit('edit', FIELD_TO_EDIT);

    expect(wrapper.findComponent(categoryFormFieldForm).exists()).toBe(true);

    expect(
      wrapper.findComponent<DefineComponent<{ categoryField: ICategoryField }>>(categoryFormFieldForm).vm.$props
        .categoryField
    ).toEqual(FIELD_TO_EDIT);
  });

  it('shows field form by add category field button click', async () => {
    expect(wrapper.findComponent(categoryFormFieldForm).exists()).toBe(false);

    await wrapper.findComponent(categoryFormAddCategory).trigger('click');

    expect(wrapper.findComponent(categoryFormFieldForm).exists()).toBe(true);
    expect(wrapper.findComponent(categoryFormAddCategory).exists()).toBe(false);
  });

  it('hides field form by hide emit', async () => {
    await wrapper.findComponent(categoryFormAddCategory).trigger('click');

    expect(wrapper.findComponent(categoryFormFieldForm).exists()).toBe(true);

    await wrapper.findComponent<DefineComponent>(categoryFormFieldForm).vm.$emit('hide');

    expect(wrapper.findComponent(categoryFormFieldForm).exists()).toBe(false);
  });

  it('updates category', async () => {
    expect(spyMutateUpdate).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);

    const NEW_CATEGORY_TITLE = 'yo';

    await wrapper.findComponent(categoryFormTitle).setValue(NEW_CATEGORY_TITLE);

    await wrapper.find(categoryForm).trigger('submit');

    expect(spyMutateUpdate).toBeCalledTimes(1);
    expect(spyMutateUpdate).toBeCalledWith({ ...CATEGORY.data, title: NEW_CATEGORY_TITLE });

    onSuccessUpdate();

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);
  });

  it('deletes banner', async () => {
    expect(spyMutateDelete).toBeCalledTimes(0);
    expect(spyRemoveQueries).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyRouterPush).toBeCalledTimes(0);

    wrapper.findComponent<DefineComponent>(categoryFormButtons).vm.$emit('delete', CATEGORY.data._id);

    expect(spyMutateDelete).toBeCalledTimes(1);
    expect(spyMutateDelete).toBeCalledWith(CATEGORY.data._id);

    onSuccessDelete();

    expect(spyRemoveQueries).toBeCalledTimes(1);
    expect(spyRemoveQueries).toBeCalledWith(QUERY_KEY);

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_CATEGORY);
  });

  it('creates category', async () => {
    wrapper.unmount();

    expect(spyMutatePost).toBeCalledTimes(0);
    expect(spyRefetchQueries).toBeCalledTimes(0);
    expect(spyToastSuccess).toBeCalledTimes(0);
    expect(spyRouterPush).toBeCalledTimes(0);

    const wrapperWithoutCategory = wrapperFactory(CategoryForm, {});

    const NEW_CATEGORY = {
      title: 'hi',
      description: 'text',
      iconUrl: 'pic.jpg',
      fields: [
        {
          _id: '6508498bdd39fcd0be367e4e',
          title: 'Base clock',
          fieldType: 'number',
          fieldValue: '',
          fieldUnits: 'GHz',
        },
      ],
    };

    await wrapperWithoutCategory.findComponent(categoryFormTitle).setValue(NEW_CATEGORY.title);
    await wrapperWithoutCategory.findComponent(categoryFormDescription).setValue(NEW_CATEGORY.description);

    wrapperWithoutCategory.vm.formData.iconUrl = NEW_CATEGORY.iconUrl;
    wrapperWithoutCategory.vm.formData.fields = [...NEW_CATEGORY.fields];

    await wrapperWithoutCategory.find(categoryForm).trigger('submit');

    expect(spyMutatePost).toBeCalledTimes(1);
    expect(spyMutatePost).toBeCalledWith(NEW_CATEGORY);

    onSuccessPost();

    expect(spyRefetchQueries).toBeCalledTimes(1);
    expect(spyRefetchQueries).toBeCalledWith(QUERY_KEY);

    await flushPromises();

    expect(spyToastSuccess).toBeCalledTimes(1);

    expect(spyRouterPush).toBeCalledTimes(1);
    expect(spyRouterPush).toBeCalledWith(URL_CATEGORY);
  });
});
