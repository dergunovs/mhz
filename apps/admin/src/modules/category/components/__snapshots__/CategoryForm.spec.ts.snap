// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CategoryForm > matches snapshot 1`] = `
"<form class="_form_93b7a2" data-test="category-form">
  <fieldset label="Title" isrequired=""><input type="text" modelvalue="CPU" isfocus="" data-test="category-form-title"></fieldset>
  <fieldset label="Description" isrequired=""><input type="text" modelvalue="<p>A <strong>central processing unit</strong> (<strong>CPU</strong>)—also called a <strong>central processor</strong> or <strong>main processor</strong>—is the most important processor in a given computer. Its electronic circuitry executes instructions of a computer program, such as arithmetic, logic, controlling, and input/output (I/O) operations. This role contrasts with that of external components, such as main memory and I/O circuitry, and specialized coprocessors such as graphics processing units (GPUs).</p>" data-test="category-form-description"></fieldset>
  <category-field-list-stub fields="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" isshowcategoryfieldform="false" data-test="category-form-field-list"></category-field-list-stub><button layout="secondary" data-test="category-form-add-category"> Add Category Field </button>
  <!--v-if--><input type="file" label="Upload icon" issingle="" isrequired="" isdisabled="true" extensions="svg" lang="en">
  <image-preview-stub urls="1695198204460-cpu.svg" isthumb="false" data-test="category-form-icon"></image-preview-stub>
  <form-buttons-stub id="64d4dc15f93f056da17ceff8" data-test="category-form-buttons"></form-buttons-stub>
</form>"
`;
